### PR TITLE
Optimize MemoryPool

### PR DESCRIPTION
**Pull request by erikzhang**

### CODE REVIEW

**Review by vncoelho — DISMISSED** [review]

File: src/neo/Ledger/MemoryPool.cs [hunk 1]
```diff
@@ -361,6 +361,11 @@ internal void UpdatePoolForBlockPersisted(Block block, DataCache snapshot)
                 _txRwLock.ExitWriteLock();
             }
 
+            // If we know about headers of future blocks, no point in verifying transactions from the unverified tx pool
+            // until we get caught up.
+            if (block.Index > 0 && Blockchain.Singleton.HeaderCache.Count > 0)
```
Comment (shargon) [inline src/neo/Ledger/MemoryPool.cs:366]: we should check the last index instead of the count, isn't it?

Comment (erikzhang) [reply src/neo/Ledger/MemoryPool.cs:366]: They are equivalent.

Comment (shargon) [reply src/neo/Ledger/MemoryPool.cs:366]: But we have a limit of 10K headers

Comment (erikzhang) [reply src/neo/Ledger/MemoryPool.cs:366]: What do you mean?

Comment (erikzhang) [reply src/neo/Ledger/MemoryPool.cs:366]: `Blockchain.Singleton.HeaderCache.Count > 0` is equivalent to `height < headerHeight`.

Comment (shargon) [reply src/neo/Ledger/MemoryPool.cs:366]: `block.Index > 0` it's required?

Comment (erikzhang) [reply src/neo/Ledger/MemoryPool.cs:366]: It's not required.

Comment (shargon) [inline src/neo/Ledger/MemoryPool.cs:366]:
```suggestion
            if (Blockchain.Singleton.HeaderCache.Count > 0)
```

**Review by shargon — APPROVED** [review]

File: src/neo/Ledger/MemoryPool.cs (hunk 1 above)
Comment (erikzhang) [reply src/neo/Ledger/MemoryPool.cs:366]: We need it. Because when `block.Index == 0`, `Blockchain.Singleton` is not available.

**Pull request merged**